src/backend/distributed/commands/view.c

```diff
@@ -0,0 +1,349 @@
+ * views will still be dropped locally but not on the workers.
+ */
+List *
+PreprocessDropViewStmt(Node *node, const char *queryString, ProcessUtilityContext
+					   processUtilityContext)
+{
+	DropStmt *stmt = castNode(DropStmt, node);
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	/*
+	 * Our statements need to be fully qualified so we can drop them from the right schema
+	 * on the workers
+	 */
+	QualifyTreeNode((Node *) stmt);
+
+	List *distributedViews = FilterNameListForDistributedViews(stmt->objects,
+															   stmt->missing_ok);
+
+	if (list_length(distributedViews) < 1)
+	{
+		/* no distributed view to drop */
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	/*
+	 * Swap the list of objects before deparsing and restore the old list after. This
+	 * ensures we only have distributed views in the deparsed drop statement.
+	 */
+	DropStmt *stmtCopy = copyObject(stmt);
+	stmtCopy->objects = distributedViews;
+	const char *dropStmtSql = DeparseTreeNode((Node *) stmtCopy);
+
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) dropStmtSql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * FilterNameListForDistributedViews takes a list of view names and filters against the
+ * views that are distributed. Given list of view names must be qualified before.
+ *
+ * The original list will not be touched, a new list will be created with only the objects
+ * in there.
+ */
+static List *
+FilterNameListForDistributedViews(List *viewNamesList, bool missing_ok)
+{
+	List *distributedViewNames = NIL;
+
+	List *qualifiedViewName = NULL;
+	foreach_ptr(qualifiedViewName, viewNamesList)
+	{
+		/*
+		 * Name of the view must be qualified before calling this function
+		 */
+		Assert(list_length(qualifiedViewName) == 2);
+
+		char *schemaName = strVal(linitial(qualifiedViewName));
+		char *viewName = strVal(lsecond(qualifiedViewName));
+
+		Oid schemaId = get_namespace_oid(schemaName, missing_ok);
+		Oid viewOid = get_relname_relid(viewName, schemaId);
+
+		if (!OidIsValid(viewOid))
+		{
+			continue;
+		}
+
+		ObjectAddress viewAddress = { 0 };
+		ObjectAddressSet(viewAddress, RelationRelationId, viewOid);
+
+		if (IsObjectDistributed(&viewAddress))
+		{
+			distributedViewNames = lappend(distributedViewNames, qualifiedViewName);
+		}
+	}
+	return distributedViewNames;
+}
+
+
+/*
+ * CreateViewDDLCommandsIdempotent returns a list of DDL statements (const char *) to be
+ * executed on a node to recreate the view addressed by the viewAddress.
+ */
+List *
+CreateViewDDLCommandsIdempotent(const ObjectAddress *viewAddress)
+{
+	StringInfo createViewCommand = makeStringInfo();
+
+	Oid viewOid = viewAddress->objectId;
+
+	char *viewName = get_rel_name(viewOid);
+	Oid schemaOid = get_rel_namespace(viewOid);
+	char *schemaName = get_namespace_name(schemaOid);
+
+	appendStringInfoString(createViewCommand, "CREATE OR REPLACE VIEW ");
```

Comment:
Sometimes `CREATE OR REPLACE VIEW` is not idempotent. (a) have you checked what kind of such cases exists -- one is the following. (b) Can we ever call this function on such views?

```SQL
create or replace view vr as select 1,* from real_city ;
ERROR:  cannot change name of view column "pop" to "?column?"
HINT:  Use ALTER VIEW ... RENAME COLUMN ... to change name of view column instead.
```

Reply:
Since PG catches such cases on the coordinator, I haven't added such tests. But let me add one to make sure it works.